pinot-core/src/main/java/org/apache/pinot/core/operator/transform/TransformBlockDataFetcher.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.segment.index.readers.Dictionary;
+
+public class TransformBlockDataFetcher {
+
+
+  private final Fetcher[] _fetchers;
+
+  public TransformBlockDataFetcher(BlockValSet[] blockValSets, Dictionary[] dictionaries,
+      TransformResultMetadata[] expressionResultMetadata) {
+    _fetchers = new Fetcher[blockValSets.length];
+    for (int i = 0; i < blockValSets.length; i++) {
+      _fetchers[i] = createFetcher(blockValSets[i], dictionaries[i],
+          expressionResultMetadata[i]);
+    }
+  }
+
+  public Serializable[] getRow(int docId) {
+    Serializable[] row = new Serializable[_fetchers.length];
+    for (int i = 0; i < _fetchers.length; i++) {
+      row[i] = _fetchers[i].getValue(docId);
+    }
+    return row;
+  }
+
+  Fetcher createFetcher(BlockValSet blockValSet,
+      Dictionary dictionary,
+      TransformResultMetadata expressionResultMetadata) {
+    if (expressionResultMetadata.hasDictionary()) {
+      if (expressionResultMetadata.isSingleValue()) {
+        return new DictionaryBasedSVValueFetcher(dictionary,
+            blockValSet.getDictionaryIdsSV(), expressionResultMetadata.getDataType());
+      } else {
+        switch (expressionResultMetadata.getDataType()) {
+          case INT:
+            return new DictionaryBasedMVIntValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+          case LONG:
+            return new DictionaryBasedMVLongValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+          case FLOAT:
+            return new DictionaryBasedMVFloatValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+          case DOUBLE:
+            return new DictionaryBasedMVDoubleValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+          case BOOLEAN:
+          case STRING:
+            return new DictionaryBasedMVStringValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+          case BYTES:
+            return new DictionaryBasedMVBytesValueFetcher(dictionary,
+                blockValSet.getDictionaryIdsMV());
+        }
+      }
+    } else {
+      switch (expressionResultMetadata.getDataType()) {
+        case INT:
+          return new SVIntValueFetcher(blockValSet.getIntValuesSV());
+        case LONG:
+          return new SVLongValueFetcher(blockValSet.getLongValuesSV());
+        case FLOAT:
+          return new SVFloatValueFetcher(blockValSet.getFloatValuesSV());
+        case DOUBLE:
+          return new SVDoubleValueFetcher(blockValSet.getDoubleValuesSV());
+        case BOOLEAN:
+        case STRING:
+          return new SVStringValueFetcher(blockValSet.getStringValuesSV());
+        case BYTES:
+          return new SVBytesValueFetcher(blockValSet.getBytesValuesSV());
+      }
+    }
+    throw new UnsupportedOperationException();
+  }
+}
+
+
+interface Fetcher {
```

Comment:
Is there a reason why we cannot use the DataFetcher that already exists?

Reply:
No, DataFetcher works on top of DataSources. Thought about refactoring the existing one to make it work for both use cases but we can do this in another PR.